apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/SystemMetrics.java

```diff
@@ -116,12 +131,44 @@ public double get() {
             }
         });
 
-        metricRegistry.addUnlessNan("system.memory.actual.free", Collections.<String, String>emptyMap(), new DoubleSupplier() {
-            @Override
-            public double get() {
-                return invoke(freeMemory);
-            }
-        });
+        if (meminfo.canRead()) {
+            metricRegistry.addUnlessNan("system.memory.actual.free", Collections.<String, String>emptyMap(), new DoubleSupplier() {
+                final List<WildcardMatcher> relevantLines = Arrays.asList(
+                    caseSensitiveMatcher("MemAvailable:*kB"),
+                    caseSensitiveMatcher("MemFree:*kB"),
+                    caseSensitiveMatcher("Buffers:*kB"),
+                    caseSensitiveMatcher("Cached:*kB"));
+
+                @Override
+                public double get() {
+                    Map<String, Long> memInfo = new HashMap<>();
```

Comment:
All this happens on the reporter thread. We need to reconsider this, not as part of this PR...